SU-SWS/ansible-playbooks
# SITES-498 Transform paths in CSS Injector files

# READY FOR REVIEW!!!!

# Summary
- Ostensibly for transforming paths in CSS Injector files. In actuality, it's a full audit and a bit of re-work of the existing find-and-replace logic.

# Needed By (Date)
- Well it's not going to be by the end of the sprint that ends today. So I guess before we migrate any more sites.

# Criticality
- How critical is this PR on a 1-10 scale? 8
- It's pretty big

# Steps to Test

1. Check out this branch
2. Use `sws-migration-testing vhost="sws-oddball-vhost"` in an inventory
3. Optionally add other GSE sites that we had issues with. TELOS, gsediversity, edarchive, crceducation?
4. Run the `migration-playbook.yml` to the `test` environment
5. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/ (your site for comparison is https://sws-oddball-vhost.stanford.edu/)
6. Verify links and images on homepage have the correct paths set
7. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/css
8. Verify that all 8 divs have a background image
9. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/admin/config/development/css-injector/edit/2
10. Verify that all paths in CSS Injector rules have been replaced correctly
11. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/view
12. Verify that all links are correct.
    1. *Exception*: Relative links to `href="/sws-migration-test/"` will not be replaced, because we are [skipping replacement of those if the vhost is defined](https://github.com/SU-SWS/ansible-playbooks/blob/SITES-498-css-injector/roles/change-paths/tasks/main.yml#L125).
    2. Ratify the logic of that decision
13. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/admin/structure/views/view/view/edit/ and verify link replacement in the header and footer
14. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/js
15. Grant yourself 4 staplers
16. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/admin/config/development/js-injector/list/js_migration_test/edit
17. Verify link replacement (for single and double-quoted strings!!)
18. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/events/sarl
19. Verify that links all are correct
20. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/admin/structure/menu/manage/main-menu
21. Verify all of the links in the "sarm" Main Menu item
22. Take a walk through the changes to the code and the (extensive comments). This part of the migration depends heavily on order-of-operations, and I have moved some things around a bit.
23. Review my comments on this PR
24. Review the tags in the `migration-playbook.yml`
25. Celebrate 🎉 

# Affected Projects or Products
- Does this PR impact any particular projects, products, or modules? Sites 2.0

# Associated Issues and/or People
- JIRA ticket: https://stanfordits.atlassian.net/browse/SITES-498 (and a few others, I'm sure)
- Other PRs:
    - #29 
    - #33 
- Anyone who should be notified? ( @sherakama )

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
